python/demo/gsoc2022/demo_sbc.py

```diff
@@ -0,0 +1,244 @@
+
+MPI.COMM_WORLD.barrier()
+
+# definition of finite element for the electric field
+curl_el = ufl.FiniteElement("N1curl", mesh.ufl_cell(), 3)
+
+# definition of finite element for the r function (see next)
+lagr_el = ufl.FiniteElement("CG", mesh.ufl_cell(), 2)
+
+# function space for the electric field
+V = fem.FunctionSpace(mesh, curl_el)
+
+# wavevector of the background field
+k0 = 2 * np.pi / wl0
+
+# angle of incidence of the background field with respect to the horizontal axis
+deg = np.pi / 180
+theta = 45 * deg
+
+# plane wave function
+f = background_electric_field(theta, n_bkg, k0)
+Eb = fem.Function(V, dtype=np.complex128)
+Eb.interpolate(f.eval)
+
+# function r = radial distance from the (0, 0) point
+lagr_space = fem.FunctionSpace(mesh, lagr_el)
+r = fem.Function(lagr_space, dtype=np.complex128)
+r.interpolate(radial_distance)
+
+# definition of Trial and Test functions
+Es = ufl.TrialFunction(V)
+Vs = ufl.TestFunction(V)
+
+# definition of 3d fields for cross and curl operations
+Es_3d = from_2d_to_3d(Es)
+Vs_3d = from_2d_to_3d(Vs)
+
+# Measures for subdomains
+dAu = ufl.Measure("dx", mesh, subdomain_data=cell_tags, subdomain_id=au_tag)
+dBkg = ufl.Measure("dx", mesh, subdomain_data=cell_tags, subdomain_id=bkg_tag)
+dsbc = ufl.Measure("ds", mesh, subdomain_data=facet_tags, subdomain_id=boundary_tag)
+dDom = dAu + dBkg
+
+# normal to the boundary
+n = FacetNormal(mesh)
+n_3d = from_2d_to_3d(n)
+
+# definition of relative permittivity for Au @400nm
+reps_au = -1.0782
+ieps_au = 5.8089
+eps_au = reps_au + ieps_au * 1j
+
+# definition of the relative permittivity over the whole domain
+D = fem.FunctionSpace(mesh, ("DG", 0))
+eps = fem.Function(D)
+au_cells = cell_tags.indices[cell_tags.values == au_tag]
+bkg_cells = cell_tags.indices[cell_tags.values == bkg_tag]
+eps.x.array[au_cells] = np.full(len(au_cells), reps_au + ieps_au * 1j)
+eps.x.array[bkg_cells] = np.full(len(bkg_cells), eps_bkg)
+
+# weak form
+F = - inner(curl(Es), curl(Vs)) * dDom \
+    + eps * k0**2 * inner(Es, Vs) * dDom \
+    + k0**2 * (eps - eps_bkg) * inner(Eb, Vs) * dDom \
+    + (1j * k0 * n_bkg + 1 / (2 * r)) * inner(cross(Es_3d, n_3d),
+                                              cross(Vs_3d, n_3d)) * dsbc  # scattering boundary condition
+
+# splitting in left-hand side and right-hand side
+a, L = lhs(F), rhs(F)
+
+problem = fem.petsc.LinearProblem(a, L, bcs=[], petsc_options={"ksp_type": "preonly", "pc_type": "lu"})
+Eh = problem.solve()
+
+# total electric field E = Es + Eb
+E = fem.Function(V, dtype=np.complex128)
+E.x.array[:] = Eb.x.array[:] + Eh.x.array[:]
+
+# ||E||
+norm_func = sqrt(inner(Eh, Eh))
+V_normEh = fem.FunctionSpace(mesh, lagr_el)
+norm_expr = fem.Expression(norm_func, V_normEh.element.interpolation_points)
+normEh = fem.Function(V_normEh)
+normEh.interpolate(norm_expr)
+
+# save the fields as xdmf files
+save_as_xdmf("data/Es.xdmf", mesh, Eh)
+save_as_xdmf("data/E.xdmf", mesh, E)
+save_as_xdmf("data/normEh.xdmf", mesh, normEh)
```

Comment:
You should not use the `XDMFFile` for writing out other functions spaces than `CG`-1.

What I recommend doing is to interpolate `Eh` into a suitable `DG` space and use `dolfinx.io.VTXWriter` to write out the functions